In [13]:
import json
import numpy as np
import pandas as pd
from sqlalchemy import create_engine

from langchain_openai import ChatOpenAI
from langchain_community.agent_toolkits import create_sql_agent
from langchain_community.utilities import SQLDatabase

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 1000)

## Import data

This data set is the same data that was used by Paco Nathan in his [blog post](https://neo4j.com/developer-blog/entity-resolved-knowledge-graphs/) on entity-resolved knowledge graphs.  It is a combination of 3 files:

1. SafeGraph: a dataset of places of interest such as businesses and organizations
2. US Department of Labor Wage and Hour Compliance Action Data (DOL_WHISARD): a publicly-available dataset of labor violations of various companies
3. US Small Business Administration PPP Loans over $150K (PPP_LOANS): a publicly-available data set of Paycheck Protection Program (PPP) was a federal loans awarded to help small businesses keep their employees during the COVID-19 pandemic

These files are available for download [here](https://senzing.com/get-erkg-tutorial-evaluation-data/).

In [5]:
sg_data = []

with open('./SafeGraph_LasVegas_Partial.json') as f:
    for line in f:
        sg_data.append(json.loads(line))

sg_df = pd.DataFrame(sg_data)

dol_data = []

with open('./Dept_Labor_Whisard_LasVegas.json') as f:
    for line in f:
        dol_data.append(json.loads(line))

dol_df = pd.DataFrame(dol_data)
dol_df = dol_df.iloc[:, :13]  # Each individual infraction count is shown beyond column 13

ppp_data = []

with open('./PPP_Loans_Over_150k_LasVegas.json') as f:
    for line in f:
        ppp_data.append(json.loads(line))

ppp_df = pd.DataFrame(ppp_data)

dol_df.replace('null', np.nan, inplace=True)

sg_df.shape, dol_df.shape, ppp_df.shape

((79946, 25), (1554, 13), (3488, 19))

## Querying tabular data

I had originally tried to use the Pandas agent from LangChain but did not have much luck with it.  So I therefore elected to stand up a quickie little SQLite database.  My brief experiment here led me to conclude that the SQL agent for LangChain was a bit more reliable.

In [ ]:
engine = create_engine("sqlite:///erkg_demo.db")
sg_df.to_sql("sg", engine, index=False)
ppp_df.to_sql("ppp", engine, index=False)
dol_df.to_sql("dol", engine, index=False)

### Verifying the SQL connection with LangChain...

In [9]:
db = SQLDatabase(engine=engine)

print(db.dialect)
print(db.get_usable_table_names())
print(db.run("SELECT count(*) FROM sg;"))
print(db.run("SELECT count(*) FROM ppp;"))
print(db.run("SELECT count(*) FROM dol;"))

sqlite
['dol', 'ppp', 'sg']
[(79946,)]
[(3488,)]
[(1554,)]


In [14]:
sql_agent_executor = create_sql_agent(
    ChatOpenAI(temperature=0, model="gpt-4o-2024-05-13"), 
    db=db, 
    agent_type="openai-tools", 
    verbose=True
)

## Let's get to work asking some questions!

First, let's just see that the SQL agent can handle some basic data ops questions.

In [15]:
sql_agent_executor.invoke('what tables are present?')



> Entering new SQL Agent Executor chain...

Invoking: `sql_db_list_tables` with `{}`


dol, ppp, sgThe database contains the following tables:
1. dol
2. ppp
3. sg

If you have any specific questions about these tables or need information from them, please let me know!

> Finished chain.


{'input': 'what tables are present?',
 'output': 'The database contains the following tables:\n1. dol\n2. ppp\n3. sg\n\nIf you have any specific questions about these tables or need information from them, please let me know!'}

In [16]:
sql_agent_executor.invoke('tell me about the tables in the database')



> Entering new SQL Agent Executor chain...

Invoking: `sql_db_list_tables` with `{}`


dol, ppp, sg
Invoking: `sql_db_schema` with `{'table_names': 'dol, ppp, sg'}`



CREATE TABLE dol (
	"RECORD_TYPE" TEXT, 
	"DATA_SOURCE" TEXT, 
	"RECORD_ID" BIGINT, 
	case_id BIGINT, 
	"BUSINESS_NAME_ORG" TEXT, 
	"LEGAL_NAME_ORG" TEXT, 
	"BUSINESS_ADDR_LINE1" TEXT, 
	"BUSINESS_ADDR_CITY" TEXT, 
	"BUSINESS_ADDR_STATE" TEXT, 
	"BUSINESS_ADDR_POSTAL_CODE" TEXT, 
	naic_cd TEXT, 
	naics_code_description TEXT, 
	case_violtn_cnt BIGINT
)

/*
3 rows from dol table:
RECORD_TYPE	DATA_SOURCE	RECORD_ID	case_id	BUSINESS_NAME_ORG	LEGAL_NAME_ORG	BUSINESS_ADDR_LINE1	BUSINESS_ADDR_CITY	BUSINESS_ADDR_STATE	BUSINESS_ADDR_POSTAL_CODE	naic_cd	naics_code_description	case_violtn_cnt
ORGANIZATION	DoL_WHISARD	53	1658108	Fabulous Freddy's (Trailwood)	Fabulous LLC	9611 Trail Wood Drive	Las Vegas	NV	89134	811192	Car Washes	0
ORGANIZATION	DoL_WHISARD	165	1419689	Boulder Station Hotel & Casino	Boulder Station, Inc.	4111 Boulder

{'input': 'tell me about the tables in the database',
 'output': 'The database contains the following tables:\n\n1. **dol**: This table appears to contain information about organizations, including their business names, addresses, NAICS codes, and violation counts.\n   - Example columns: `RECORD_TYPE`, `DATA_SOURCE`, `RECORD_ID`, `BUSINESS_NAME_ORG`, `BUSINESS_ADDR_CITY`, `BUSINESS_ADDR_STATE`, `naics_code_description`, `case_violtn_cnt`.\n\n2. **ppp**: This table seems to store data related to PPP loans, including loan ranges, business names, addresses, NAICS codes, business types, and other related information.\n   - Example columns: `RECORD_TYPE`, `DATA_SOURCE`, `RECORD_ID`, `Loan_Range`, `BUSINESS_NAME_ORG`, `BUSINESS_ADDR_CITY`, `BUSINESS_ADDR_STATE`, `NAICS_Code`, `Business_Type`, `JobsReported`, `DateApproved`, `Lender`.\n\n3. **sg**: This table contains information about various locations, including their names, brands, categories, NAICS codes, geographical coordinates, and oth

## Now let's ask some questions that start looking at the data itself...

In [17]:
sql_agent_executor.invoke('find all references to Mandalay Bay in all of the tables')



> Entering new SQL Agent Executor chain...

Invoking: `sql_db_list_tables` with `{}`


dol, ppp, sg
Invoking: `sql_db_schema` with `{'table_names': 'dol'}`



CREATE TABLE dol (
	"RECORD_TYPE" TEXT, 
	"DATA_SOURCE" TEXT, 
	"RECORD_ID" BIGINT, 
	case_id BIGINT, 
	"BUSINESS_NAME_ORG" TEXT, 
	"LEGAL_NAME_ORG" TEXT, 
	"BUSINESS_ADDR_LINE1" TEXT, 
	"BUSINESS_ADDR_CITY" TEXT, 
	"BUSINESS_ADDR_STATE" TEXT, 
	"BUSINESS_ADDR_POSTAL_CODE" TEXT, 
	naic_cd TEXT, 
	naics_code_description TEXT, 
	case_violtn_cnt BIGINT
)

/*
3 rows from dol table:
RECORD_TYPE	DATA_SOURCE	RECORD_ID	case_id	BUSINESS_NAME_ORG	LEGAL_NAME_ORG	BUSINESS_ADDR_LINE1	BUSINESS_ADDR_CITY	BUSINESS_ADDR_STATE	BUSINESS_ADDR_POSTAL_CODE	naic_cd	naics_code_description	case_violtn_cnt
ORGANIZATION	DoL_WHISARD	53	1658108	Fabulous Freddy's (Trailwood)	Fabulous LLC	9611 Trail Wood Drive	Las Vegas	NV	89134	811192	Car Washes	0
ORGANIZATION	DoL_WHISARD	165	1419689	Boulder Station Hotel & Casino	Boulder Station, Inc.	4111 Boulder Hwy	Las 

{'input': 'find all references to Mandalay Bay in all of the tables',
 'output': 'Here are the references to "Mandalay Bay" found in the `sg` table:\n\n1. **Location Name:** Mandalay Bay Sports Book\n   - **Address:** 3950 Las Vegas Blvd S, Las Vegas, NV 89119-1005\n\n2. **Location Name:** The Gym At Mandalay Bay\n   - **Address:** 3930 Las Vegas Blvd S, Las Vegas, NV 89119-1010\n\n3. **Location Name:** Mandalay Bay Resort & Casino\n   - **Address:** 3960 Las Vegas Blvd S, Las Vegas, NV 89119-1046\n\n4. **Location Name:** Mandalay Bay Self Parking\n   - **Address:** 3930 Las Vegas Blvd S, Las Vegas, NV 89119-1010\n\n5. **Location Name:** Mandalay Bay Parking Garage\n   - **Address:** Not available\n\n6. **Location Name:** Kalologie Medspa at Mandalay Bay\n   - **Address:** 3950 Las Vegas Blvd S, Las Vegas, NV 89119-1005\n\n7. **Location Name:** Regional Transportation Commission of Southern Nevada SB Las Vegas at Mandalay Bay\n   - **Address:** Not available\n\n8. **Location Name:** Ho

In [18]:
sql_agent_executor.invoke('does Mandalay Bay have any violations?')



> Entering new SQL Agent Executor chain...

Invoking: `sql_db_list_tables` with `{}`


dol, ppp, sg
Invoking: `sql_db_schema` with `{'table_names': 'dol, ppp, sg'}`



CREATE TABLE dol (
	"RECORD_TYPE" TEXT, 
	"DATA_SOURCE" TEXT, 
	"RECORD_ID" BIGINT, 
	case_id BIGINT, 
	"BUSINESS_NAME_ORG" TEXT, 
	"LEGAL_NAME_ORG" TEXT, 
	"BUSINESS_ADDR_LINE1" TEXT, 
	"BUSINESS_ADDR_CITY" TEXT, 
	"BUSINESS_ADDR_STATE" TEXT, 
	"BUSINESS_ADDR_POSTAL_CODE" TEXT, 
	naic_cd TEXT, 
	naics_code_description TEXT, 
	case_violtn_cnt BIGINT
)

/*
3 rows from dol table:
RECORD_TYPE	DATA_SOURCE	RECORD_ID	case_id	BUSINESS_NAME_ORG	LEGAL_NAME_ORG	BUSINESS_ADDR_LINE1	BUSINESS_ADDR_CITY	BUSINESS_ADDR_STATE	BUSINESS_ADDR_POSTAL_CODE	naic_cd	naics_code_description	case_violtn_cnt
ORGANIZATION	DoL_WHISARD	53	1658108	Fabulous Freddy's (Trailwood)	Fabulous LLC	9611 Trail Wood Drive	Las Vegas	NV	89134	811192	Car Washes	0
ORGANIZATION	DoL_WHISARD	165	1419689	Boulder Station Hotel & Casino	Boulder Station, Inc.	4111 Boulder

{'input': 'does Mandalay Bay have any violations?',
 'output': 'Yes, Mandalay Bay has violations. Here are the details:\n\n- One record shows 2 violations.\n- Two records show 0 violations.'}

In [19]:
sql_agent_executor.invoke('which company has the most violations?')



> Entering new SQL Agent Executor chain...

Invoking: `sql_db_list_tables` with `{}`


dol, ppp, sg
Invoking: `sql_db_schema` with `{'table_names': 'dol'}`



CREATE TABLE dol (
	"RECORD_TYPE" TEXT, 
	"DATA_SOURCE" TEXT, 
	"RECORD_ID" BIGINT, 
	case_id BIGINT, 
	"BUSINESS_NAME_ORG" TEXT, 
	"LEGAL_NAME_ORG" TEXT, 
	"BUSINESS_ADDR_LINE1" TEXT, 
	"BUSINESS_ADDR_CITY" TEXT, 
	"BUSINESS_ADDR_STATE" TEXT, 
	"BUSINESS_ADDR_POSTAL_CODE" TEXT, 
	naic_cd TEXT, 
	naics_code_description TEXT, 
	case_violtn_cnt BIGINT
)

/*
3 rows from dol table:
RECORD_TYPE	DATA_SOURCE	RECORD_ID	case_id	BUSINESS_NAME_ORG	LEGAL_NAME_ORG	BUSINESS_ADDR_LINE1	BUSINESS_ADDR_CITY	BUSINESS_ADDR_STATE	BUSINESS_ADDR_POSTAL_CODE	naic_cd	naics_code_description	case_violtn_cnt
ORGANIZATION	DoL_WHISARD	53	1658108	Fabulous Freddy's (Trailwood)	Fabulous LLC	9611 Trail Wood Drive	Las Vegas	NV	89134	811192	Car Washes	0
ORGANIZATION	DoL_WHISARD	165	1419689	Boulder Station Hotel & Casino	Boulder Station, Inc.	4111 Boulder Hwy	Las 

{'input': 'which company has the most violations?',
 'output': 'The company with the most violations is **MGM Grand Hotel & Casino** with a total of 16,468 violations. Here are the top 10 companies with the most violations:\n\n1. MGM Grand Hotel & Casino - 16,468 violations\n2. Marquis Care Plaza Regency - 12,000 violations\n3. Suncoast Hotel & Casino - 3,273 violations\n4. C Martin Company Inc - 1,177 violations\n5. Desert Plastering, LLC - 1,060 violations\n6. One Stop Motors.Com - 1,026 violations\n7. WestCor Construction - 933 violations\n8. Garris Plastering - 723 violations\n9. Checker Cab Company - 685 violations\n10. Westgate Marketing - 680 violations'}

## Uh oh!

Houston, we have a problem!

According to these basic SQL queries, Mandalay Bay has only a few violations.  While that is true for these very specific rows, it misses the fact that Mandalay Bay is owned by MGM Grand, which has the most violations!  So if you were wanting to get an accurate accounting of those violations (or many other questions you could think of to ask this data), you will miss it because you have not resolved the fact of the parent relationship between MGM Grand and Mandalay Bay!

### Now let's ask something a little more complicated...

In [21]:
sql_agent_executor.invoke('of the companies that took on the biggest loans, which had the most violations?')



> Entering new SQL Agent Executor chain...

Invoking: `sql_db_list_tables` with `{}`


dol, ppp, sg
Invoking: `sql_db_schema` with `{'table_names': 'dol'}`



CREATE TABLE dol (
	"RECORD_TYPE" TEXT, 
	"DATA_SOURCE" TEXT, 
	"RECORD_ID" BIGINT, 
	case_id BIGINT, 
	"BUSINESS_NAME_ORG" TEXT, 
	"LEGAL_NAME_ORG" TEXT, 
	"BUSINESS_ADDR_LINE1" TEXT, 
	"BUSINESS_ADDR_CITY" TEXT, 
	"BUSINESS_ADDR_STATE" TEXT, 
	"BUSINESS_ADDR_POSTAL_CODE" TEXT, 
	naic_cd TEXT, 
	naics_code_description TEXT, 
	case_violtn_cnt BIGINT
)

/*
3 rows from dol table:
RECORD_TYPE	DATA_SOURCE	RECORD_ID	case_id	BUSINESS_NAME_ORG	LEGAL_NAME_ORG	BUSINESS_ADDR_LINE1	BUSINESS_ADDR_CITY	BUSINESS_ADDR_STATE	BUSINESS_ADDR_POSTAL_CODE	naic_cd	naics_code_description	case_violtn_cnt
ORGANIZATION	DoL_WHISARD	53	1658108	Fabulous Freddy's (Trailwood)	Fabulous LLC	9611 Trail Wood Drive	Las Vegas	NV	89134	811192	Car Washes	0
ORGANIZATION	DoL_WHISARD	165	1419689	Boulder Station Hotel & Casino	Boulder Station, Inc.	4111 Boulder Hwy	Las 

{'input': 'of the companies that took on the biggest loans, which had the most violations?',
 'output': 'The company that took on one of the biggest loans and had the most violations is HSG, LLC, with a loan range of $350,000-1 million and 36 violations.'}

### Interesting.  Who are these folks?

In [22]:
sql_agent_executor.invoke('find all references to HSG, LLC')



> Entering new SQL Agent Executor chain...

Invoking: `sql_db_list_tables` with `{}`


dol, ppp, sg
Invoking: `sql_db_schema` with `{'table_names': 'dol'}`



CREATE TABLE dol (
	"RECORD_TYPE" TEXT, 
	"DATA_SOURCE" TEXT, 
	"RECORD_ID" BIGINT, 
	case_id BIGINT, 
	"BUSINESS_NAME_ORG" TEXT, 
	"LEGAL_NAME_ORG" TEXT, 
	"BUSINESS_ADDR_LINE1" TEXT, 
	"BUSINESS_ADDR_CITY" TEXT, 
	"BUSINESS_ADDR_STATE" TEXT, 
	"BUSINESS_ADDR_POSTAL_CODE" TEXT, 
	naic_cd TEXT, 
	naics_code_description TEXT, 
	case_violtn_cnt BIGINT
)

/*
3 rows from dol table:
RECORD_TYPE	DATA_SOURCE	RECORD_ID	case_id	BUSINESS_NAME_ORG	LEGAL_NAME_ORG	BUSINESS_ADDR_LINE1	BUSINESS_ADDR_CITY	BUSINESS_ADDR_STATE	BUSINESS_ADDR_POSTAL_CODE	naic_cd	naics_code_description	case_violtn_cnt
ORGANIZATION	DoL_WHISARD	53	1658108	Fabulous Freddy's (Trailwood)	Fabulous LLC	9611 Trail Wood Drive	Las Vegas	NV	89134	811192	Car Washes	0
ORGANIZATION	DoL_WHISARD	165	1419689	Boulder Station Hotel & Casino	Boulder Station, Inc.	4111 Boulder Hwy	Las 

{'input': 'find all references to HSG, LLC',
 'output': 'Here are the references to "HSG, LLC" found in the database:\n\nFrom the `dol` table:\n- **Business Name:** HSG, LLC\n  - **Legal Name:** HSG, LLC\n  - **Address:** 4001 South Decatur Blvd, Las Vegas, NV, 89103\n\nFrom the `ppp` table:\n- **Business Name:** HSG, LLC\n  - **Address:** 4001 S. Decatur Blvd. #37-376, Las Vegas, NV, 89103\n\nNo references to "HSG, LLC" were found in the `sg` table.'}

### Some details about this

So I looked and it turns out that 4001 South Decatur Blvd is a strip mall.  And it just so happens that Suite 37 is a UPS Store, which means that this is probably a company with a PO box in that store.  In fact, if you search you will find that there are several businesses in this data set with the same address.  However, interesting things happen if you look at this entry within `G2Explorer.py`.  (See the blog post for that!)